runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowMetrics.java

```diff
@@ -206,6 +122,167 @@ public MetricQueryResults queryMetrics(MetricsFilter filter) {
+
+    public Iterable<MetricResult<GaugeResult>> getGaugeResults() {
+      return gaugeResults.build();
+    }
+  }
+
+  private static class DataflowMetricQueryResultsFactory {
+    Iterable<com.google.api.services.dataflow.model.MetricUpdate> metricUpdates;
+    MetricsFilter filter;
+    HashMap<MetricKey, com.google.api.services.dataflow.model.MetricUpdate> tentativeByName;
+    HashMap<MetricKey, com.google.api.services.dataflow.model.MetricUpdate> committedByName;
+    HashSet<MetricKey> metricHashKeys;
+    private DataflowPipelineJob dataflowPipelineJob;
+
+    public static DataflowMetricQueryResultsFactory create(DataflowPipelineJob dataflowPipelineJob,
+        Iterable<com.google.api.services.dataflow.model.MetricUpdate> metricUpdates,
+        MetricsFilter filter) {
+      return new DataflowMetricQueryResultsFactory(dataflowPipelineJob, metricUpdates, filter);
+    }
+
+    private DataflowMetricQueryResultsFactory(DataflowPipelineJob dataflowPipelineJob,
+        Iterable<com.google.api.services.dataflow.model.MetricUpdate> metricUpdates,
+        MetricsFilter filter) {
+      this.dataflowPipelineJob = dataflowPipelineJob;
+      this.metricUpdates = metricUpdates;
+      this.filter = filter;
+
+      tentativeByName = new HashMap<>();
+      committedByName = new HashMap<>();
+      metricHashKeys = new HashSet<>();
+    }
+
+    /**
+     * Check whether a {@link com.google.api.services.dataflow.model.MetricUpdate} is a tentative
+     * update or not.
+     * @return true if update is tentative, false otherwise
+     */
+    private boolean isMetricTentative(
+        com.google.api.services.dataflow.model.MetricUpdate metricUpdate) {
+      return (metricUpdate.getName().getContext().containsKey("tentative")
+          && Objects.equal(metricUpdate.getName().getContext().get("tentative"), "true"));
+    }
+
+    /**
+     * Build an immutable map that serves as a hash key for a metric update.
+     * @return a {@link MetricKey} that can be hashed and used to identify a metric.
+     */
+    private MetricKey getMetricHashKey(
+        com.google.api.services.dataflow.model.MetricUpdate metricUpdate) {
+      String fullStepName = metricUpdate.getName().getContext().get("step");
+      fullStepName = (dataflowPipelineJob.transformStepNames != null
+          ? dataflowPipelineJob.transformStepNames
+          .inverse().get(fullStepName).getFullName() : fullStepName);
+      return MetricKey.create(
+          fullStepName,
+          MetricName.named(
+              metricUpdate.getName().getContext().get("namespace"),
+              metricUpdate.getName().getName()));
+    }
+
+    private void buildMetricsIndex() {
+      // If the Context of the metric update does not have a namespace, then these are not
+      // actual metrics counters.
+      for (com.google.api.services.dataflow.model.MetricUpdate update : metricUpdates) {
+        if (Objects.equal(update.getName().getOrigin(), "user") && isMetricTentative(update)
```

Comment:
This may be clearer as:
```
if (Objects.equal(update.getName().getOrigin(), "user")
    && update.getName().getContext().containsKey("namespace")) {
  if (isMetricTentative(update) {
    tentativeByName.put(...);
  } else {
    committedByName.put(...);
  }
  metricHashKeys.add(...);
}